perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -2401,36 +2402,61 @@ public void checkGroupExists(PerunSession sess, Group group) throws GroupNotExis
 
 	@Override
 	public List<Member> filterMembersByMembershipTypeInGroup(List<Member> members) {
-		Set<Member> filteredMembers = new HashSet<>();
-		Iterator<Member> membersIterator = members.iterator();
+		Map<Integer, List<Member>> indirectMembersById = members.stream()
+				.filter(m -> m.getMembershipType().equals(MembershipType.INDIRECT))
+				.collect(groupingBy(Member::getId, toList()));
 
-		//Add members with direct membership type
-		while(membersIterator.hasNext()) {
-			Member m = membersIterator.next();
-			if(m.getMembershipType().equals(MembershipType.DIRECT)) {
-				filteredMembers.add(m);
-				membersIterator.remove();
-			}
-		}
+		List<Member> directMembers = members.stream()
+				.filter(m -> m.getMembershipType().equals(MembershipType.DIRECT))
+				.distinct()
```

Comment:
I am just curious. Is this distinct necessary?

Reply:
The old implementation used Set, for some reason, and I wasn't sure if this could happen for the direct members as well. Probably not. But, this shouldn't cause any unwanted side effects.